platform/pwl.c

```diff
@@ -1103,6 +1115,93 @@ zxdg_shell_v6_ping (void                 *data G_GNUC_UNUSED,
 #endif /* HAVE_XDG_SHELL_UNSTABLE_V6 */
 
 
+static PwlSeat*
+pwl_seat_create(struct wl_seat *seat, uint32_t name, uint32_t version, PwlDisplay *display)
+{
+    PwlSeat *self = g_slice_new0 (PwlSeat);
+    self->seat = seat;
+    self->seat_name = name;
+    self->seat_version = version;
+
+    self->display = display;
+    if (display)
+    {
+        wl_list_init (&self->link);
+        wl_list_insert (&display->seats, &self->link);
+    }
+
+    static const struct wl_seat_listener seat_listener = {
+        .capabilities = display_on_seat_capabilities,
+        .name = display_on_seat_name,
+    };
+    wl_seat_add_listener (seat, &seat_listener, self);
+
+    if (!(self->xkb_data.context = xkb_context_new (XKB_CONTEXT_NO_FLAGS))) {
+        g_error ("Could not initialize XKB context");
+        return self;
+    }
+
+    struct xkb_compose_table* compose_table =
+        xkb_compose_table_new_from_locale (self->xkb_data.context,
+                                           setlocale (LC_CTYPE, NULL),
+                                           XKB_COMPOSE_COMPILE_NO_FLAGS);
+    if (compose_table) {
+        self->xkb_data.compose_state =
+            xkb_compose_state_new (compose_table,
+                                   XKB_COMPOSE_STATE_NO_FLAGS);
+        g_clear_pointer (&compose_table, xkb_compose_table_unref);
+
+        if (!self->xkb_data.compose_state) {
+            g_error ("Could not initialize XKB compose state");
+            return self;
+        }
+    }
+
+    /* Always have a default keymap, in case we never receive .keymap. */
+    self->xkb_data.keymap =
+        xkb_keymap_new_from_names (self->xkb_data.context,
+                                   &((struct xkb_rule_names) { }),
+                                   XKB_KEYMAP_COMPILE_NO_FLAGS);
+    if (!self->xkb_data.keymap) {
+        g_error ("Could not initialize XKB keymap");
+        return self;
+    }
+
+    if (!(self->xkb_data.state = xkb_state_new (self->xkb_data.keymap))) {
+        g_error ("Could not initialize XKB state");
+        return self;
+    }
+
+    self->xkb_data.indexes.control = xkb_keymap_mod_get_index (self->xkb_data.keymap,
+                                                               XKB_MOD_NAME_CTRL);
+    self->xkb_data.indexes.alt = xkb_keymap_mod_get_index (self->xkb_data.keymap,
+                                                           XKB_MOD_NAME_ALT);
+    self->xkb_data.indexes.shift = xkb_keymap_mod_get_index (self->xkb_data.keymap,
+                                                             XKB_MOD_NAME_SHIFT);
+    return self;
+}
+
+static void
+pwl_seat_destroy(PwlSeat *self)
```

Comment:
Missing space between the function name and `(`.

Reply:
done